cpu/stm32f4: add riotboot requirements

fjmolinas opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->

This PR adds riotboot for stm32f4. Since F4 works with sectors for riotboot to work each slot must start at the beginning of a ~~slot~~sector. For stm32f4 the smallest ~~slot~~sector is of length 16kB, therefore that must be the length of the bootloader and the same amount must be taken from the second half of the flash to optimize flash usage.

This PR is part of 3, since just adding riotboot doesn't make much sense if it can't perform ota. Although it doesn't depend on them it is highly related to:

* Support flashpage: https://github.com/RIOT-OS/RIOT/pull/11681
* Don't erase fist page at un update finish: https://github.com/RIOT-OS/RIOT/pull/11683

### Testing procedure

It has been tested on the supported boards: `nucleo-f446re` and `stm32f429i-disc1`. 

1.- Test riotboot:

`BOARD=nucleo-f446r FEATURES_REQUIRED+=riotboot make -C tests/xtimer_usleep riotboot/flash-extended-slot0 term`

`BOARD=nucleo-f446rFEATURES_REQUIRED+=riotboot make -C tests/xtimer_usleep riotboot/flash-slot1 term`

In both cases it should run without issues.

2.- Since the sector division is important when updating the slots by writing to flash directly it would be ideal to reba se on top off flashpage: https://github.com/RIOT-OS/RIOT/pull/11681 and https://github.com/RIOT-OS/RIOT/pull/11683

and run `tests/riotboot_flashwrite`

The following patch can be applied to the test Makefile

```
diff --git a/tests/riotboot_flashwrite/Makefile b/tests/riotboot_flashwrite/Makefile
index 32f127d70..d3c9f20b4 100644
--- a/tests/riotboot_flashwrite/Makefile
+++ b/tests/riotboot_flashwrite/Makefile
@@ -39,11 +39,17 @@ LOW_MEMORY_BOARDS := nucleo-f334r8
 GNRC_NETIF_NUMOF := 2
 
 # uncomment these to use ethos
-#USEMODULE += ethos gnrc_uhcpc
+USEMODULE += ethos gnrc_uhcpc  stdio_uart_rx
 #
 ## ethos baudrate can be configured from make command
-#ETHOS_BAUDRATE ?= 115200
-#CFLAGS += -DETHOS_BAUDRATE=$(ETHOS_BAUDRATE) -DUSE_ETHOS_FOR_STDIO
+ETHOS_BAUDRATE ?= 115200
+CFLAGS += -DETHOS_BAUDRATE=$(ETHOS_BAUDRATE) -DUSE_ETHOS_FOR_STDIO
+
+TAP ?= tap0
+IPV6_PREFIX ?= 2001:db8::/64
+
+TERMPROG ?= sudo sh $(RIOTTOOLS)/ethos/start_network.sh
+TERMFLAGS ?= $(PORT) $(TAP) $(IPV6_PREFIX)
 
 ifneq (,$(filter $(BOARD),$(LOW_MEMORY_BOARDS)))
   $(info Using low-memory configuration for microcoap_server.)

```

Then to test:

- provided the node `make -C tests/riotboot_flashwrite/ BOARD=stm32f429i-disc1 riotboot/flash term`
- compile new firmware: `make -C tests/riotboot_flashwrite/ BOARD=stm32f429i-disc1 riotboot`
- launch an updated: `coap-client -m post coap://[fe80::2%tap0]/flashwrite -f tests/riotboot_flashwrite/bin/stm32f429i-disc1/tests_riotboot_flashwrite-slot1.riot.bin -b 64`

When the update finished reboot the node manually it should have started from a different slot.

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

Related to  https://github.com/RIOT-OS/RIOT/pull/11681 and https://github.com/RIOT-OS/RIOT/pull/11683

aabadie commented:
> Since F4 works with sectors for riotboot to work each slot must start at the beginning of a slot. For stm32f4 the smallest slot is of length 16kB

It seems there is a confusion between slots and sectors in this sentence. From my understanding, it should be: Since F4 works with sectors, for riotboot to work, each slot must start at the beginning of a **sector**. For stm32f4 the smallest **sector** is of length 16kB...

This PR also needs a rebase.

aabadie commented on cpu/stm32f4/Makefile.include:
```diff
@@ -1,5 +1,24 @@
+
+# "The Vector table must be naturally aligned to a power of two whose alignment
+# value is greater than or equal to number of Exceptions supported x 4"
+# CPU_IRQ_NUMOFF for stm32l4 boards is < 102+16 so (118*4 bytes = 472 bytes ~= 0x200)
```
Is this comment block still needed for F4? There's a mention to stm32l4, so it looks like a copy paste left-over.

aabadie commented on cpu/stm32f4/Makefile.include:
```diff
@@ -1,5 +1,24 @@
+# RIOTBOOT_HDR_LEN can be set to 0x200
+RIOTBOOT_HDR_LEN ?= 0x200
+
+# Sectors don't have the same length, they go from 16KB to 128KB. The first 5 sectors adding
```
"The sum of the 5 first sectors is up to 128kB" might be more readable (if it's what you mean here).

aabadie commented on cpu/stm32f4/Makefile.include:
```diff
@@ -1,5 +1,24 @@
+RIOTBOOT_HDR_LEN ?= 0x200
+
+# Sectors don't have the same length, they go from 16KB to 128KB. The first 5 sectors adding
+# up to 128KB. Since flash can only be erase by sector, SLOT's can't overlap. The most
```
s/be erase/be erased/

aabadie replied:
"slots can't overlap a sector" might be better ?

aabadie commented on cpu/stm32f4/Makefile.include:
```diff
@@ -1,5 +1,24 @@
+# Sectors don't have the same length, they go from 16KB to 128KB. The first 5 sectors adding
+# up to 128KB. Since flash can only be erase by sector, SLOT's can't overlap. The most
+# efficient layout comes from removing $(RIOTBOOT_LEN) twice, once for the actual bootloader
+# at the start of the flash, and a second time and the middle of the flash, for even sized slots.
```
**in** the middle of the flash

fjmolinas commented:
@aabadie comments addressed.

aabadie commented:
Please squash and trigger Murdock after.

fjmolinas commented:
@aabadie I squashed and modified the commit message, there where some typos. I also removed:

```
SLOT0_OFFSET ?= $(RIOTBOOT_LEN)	
SLOT1_OFFSET ?= $(shell echo $$(($(SLOT0_OFFSET)+ $(RIOTBOOT_LEN) + $(SLOT0_LEN))))
```

Which was already defined for riotboot.

aabadie commented:
Ok, I'll have a second look tomorrow, if I find time, just in case.

aabadie commented:
I tested a second time this PR and it works. Let's go!

fjmolinas commented:
@aabadie Thanks for the review!
